### PR TITLE
Drop Ruby 2.4/2.5/2.6 support

### PR DESCRIPTION
**Pull request by bastelfreak**

### CODE REVIEW

**Comment by bastelfreak** [comment]
To not cause a huge diff, I didn't fix rubocop violations, but updated the .rubocop_todo.yml. I would like to do some cleanup in a followup PR.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -15,20 +15,18 @@ jobs:
       - name: Install Ruby ${{ matrix.ruby }}
         uses: ruby/setup-ruby@v1
         with:
-          ruby-version: "3.1"
+          # run rubocop with the ruby version we lint for
+          ruby-version: "2.7"
```
Comment (ekohl) [inline .github/workflows/test.yml:19]: Why? Normally you set target version in RuboCop (which you did) so why does it matter?

File: Rakefile [hunk 2]
```diff
@@ -343,3 +343,15 @@ begin
 rescue LoadError
   # Optional group in bundler
 end
+
+require 'rubocop/rake_task'
+RuboCop::RakeTask.new(:rubocop) do |task|
+  # These make the rubocop experience maybe slightly less terrible
+  task.options = ['-D', '-S', '-E']
```
Comment (ekohl) [inline Rakefile:350]: Are there long form options?

File: .rubocop.yml [hunk 3]
```diff
@@ -23,7 +24,7 @@ Layout:
 Metrics:
   Enabled: false
 
-Style:
+Style/TrailingCommaInArguments:
```
Comment (ekohl) [inline .rubocop.yml:27]: This leads to a massively larger TODO file. Please keep all style changes for a separate PR.

File: Gemfile [hunk 4]
```diff
@@ -20,11 +20,11 @@ if ENV['BEAKER_PE_PR_REPO_URL']
   gem lib, :git => "git@github.com:#{author}/#{lib}.git", :branch => ref
 end
 
-group :rubocop do
-  gem 'rubocop', '~> 1.12.0'
-  gem 'rubocop-performance'
-  gem 'rubocop-rake'
-  gem 'rubocop-rspec'
+group :tests do
```
Comment (ekohl) [inline Gemfile:23]: I liked the separate rubocop group. We don't do it today, but we can exclude RuboCop from our tests this way.

File: Rakefile [hunk 5]
```diff
@@ -343,3 +343,15 @@ begin
...
+
+  # Use Rubocop's Github Actions formatter if possible
+  if ENV['GITHUB_ACTIONS'] == 'true'
+    rubocop_spec = Gem::Specification.find_by_name('rubocop')
+    task.formatters << 'github' if Gem::Version.new(rubocop_spec.version) >= Gem::Version.new('1.2')
```
Comment (ekohl) [inline Rakefile:355]: Since you require a new enough version, there's no need to try to be compatible with old versions.
```suggestion
    task.formatters << 'github'
```

File: lib/beaker/dsl/structure.rb [hunk 6]
```diff
@@ -58,7 +58,7 @@ def step step_name, &block
               else
                 logger.info("Exception raised during step execution and debug-errors option is set, entering pry. Exception was: #{e.inspect}")
                 logger.info("HINT: Use the pry 'backtrace' and 'up' commands to navigate to the test code")
-                binding.pry # rubocop:disable Lint/Debugger
```
Comment (ekohl) [inline lib/beaker/dsl/structure.rb:61]: Please keep this disable statement. We explicitly want pry, so moving it to TODO is counter-productive.

File: lib/beaker/options/parser.rb [hunk 7]
```diff
@@ -297,7 +297,7 @@ def parse_hosts_options
         hosts_file_content = begin
           bhg_cli = BeakerHostGenerator::CLI.new(host_generator_options)
           bhg_cli.execute
-        rescue BeakerHostGenerator::Exceptions::Error,
+        rescue BeakerHostGenerator::Exceptions::Error, # rubocop:disable Lint/ShadowedException
           BeakerHostGenerator::Exceptions::InvalidNodeSpecError => e
```
Comment (ekohl) [inline lib/beaker/options/parser.rb:301]: Please fix the cop instead. It's complaining (rightfully so) that `InvalidNodeSpecError` is a subclass of `Error` and thus the rescue is redundant.
```suggestion
        rescue BeakerHostGenerator::Exceptions::Error => e
```

File: Gemfile [hunk 8]
```diff
@@ -20,11 +20,11 @@ if ENV['BEAKER_PE_PR_REPO_URL']
...
+  gem 'rubocop', '~> 1.45.0'
+  gem 'rubocop-rspec', '~> 2.18.1'
+  gem 'rubocop-rake', '~> 0.6.0'
+  gem 'rubocop-performance', '~> 1.16.0'
```
Comment (ekohl) [inline Gemfile:27]: Previously the lines were sorted (and there's a cop that enforces it). Please keep that.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: Rakefile (hunk 2 above)
Comment (bastelfreak) [reply Rakefile]: updated

File: Gemfile (hunk 4 above)
Comment (bastelfreak) [reply Gemfile]: done

File: Rakefile (hunk 5 above)
Comment (bastelfreak) [reply Rakefile]: done

File: Gemfile (hunk 8 above)
Comment (bastelfreak) [reply Gemfile]: done

File: lib/beaker/options/parser.rb (hunk 7 above)
Comment (bastelfreak) [reply lib/beaker/options/parser.rb:301]: done

File: lib/beaker/dsl/structure.rb (hunk 6 above)
Comment (bastelfreak) [reply lib/beaker/dsl/structure.rb:61]: We cannot use the statement here because we disable all `Lint` rubocops in `.rubocop.yml`:

```
Lint:
  Enabled: false
```

I would like to remove that from .rubocop.yml in a followup PR and will readd the comment to lib/beaker/dsl/structure.rb.

File: .rubocop.yml [hunk 9]
```diff
@@ -2,11 +2,12 @@
 inherit_from: .rubocop_todo.yml
 
 require:
-  - rubocop-performance
   - rubocop-rake
   - rubocop-rspec
+  - rubocop-performance
```
[the remark this answers is not in the document]
Comment (bastelfreak) [reply .rubocop.yml:7]: done

File: .rubocop_todo.yml [hunk 10]
```diff
@@ -1,11 +1,40 @@
...
+# SupportedStyles: Gemfile, gems.rb, gemspec
+# Include: **/*.gemspec, **/Gemfile, **/gems.rb
+Gemspec/DevelopmentDependencies:
+  Exclude:
+    - 'beaker.gemspec'
+
+# This cop supports safe autocorrection (--autocorrect).
+# Configuration parameters: Severity, Include.
+# Include: **/*.gemspec
+Gemspec/RequireMFA:
```
[the remark this answers is not in the document]
Comment (bastelfreak) [reply .rubocop_todo.yml:26]: done

File: .rubocop_todo.yml [hunk 11]
```diff
@@ -35,11 +69,21 @@ Lint/EmptyBlock:
     - 'spec/beaker/logger_junit_spec.rb'
     - 'spec/beaker/logger_spec.rb'
 
+# This cop supports safe autocorrection (--autocorrect).
+Lint/ErbNewArguments:
```
[the remark this answers is not in the document]
Comment (bastelfreak) [reply .rubocop_todo.yml:73]: done

File: .rubocop_todo.yml [hunk 12]
```diff
@@ -1,11 +1,40 @@
...
+# using RuboCop version 1.45.1.
 # The point is for the user to remove these configuration records
 # one by one as the offenses are removed from the code base.
 # Note that changes in the inspected code, or installation of new
 # versions of RuboCop, may require this file to be generated again.
 
+# This cop supports safe autocorrection (--autocorrect).
+# Configuration parameters: Severity, Include.
+# Include: **/*.gemspec
+Gemspec/DeprecatedAttributeAssignment:
```
[the remark this answers is not in the document]
Comment (bastelfreak) [reply .rubocop_todo.yml]: done

File: lib/beaker/dsl/structure.rb (hunk 6 above)
Comment (ekohl) [reply lib/beaker/dsl/structure.rb:61]: I don't see that, and if we did then I wouldn't expect Lint to show up as a TODO item.

File: Rakefile [hunk 13]
```diff
@@ -343,3 +343,17 @@ begin
...
+    # These make the rubocop experience maybe slightly less terrible
+    task.options = ['--display-cop-names', '--display-style-guide', '--extra-details']
+
+    # Use Rubocop's Github Actions formatter if possible
+    if ENV['GITHUB_ACTIONS'] == 'true'
+      task.formatters << 'github'
+    end
+  end
+rescue LoadError
+end
```
Comment (ekohl) [inline Rakefile:359]: I think you're introducing a RuboCop violation by not having a comment (nor code) in the rescue block. I always like to write it like this:

```suggestion
begin
  require 'rubocop/rake_task'
rescue LoadError
  # RuboCop is an optional group
else
  RuboCop::RakeTask.new(:rubocop) do |task|
    # These make the rubocop experience maybe slightly less terrible
    task.options = ['--display-cop-names', '--display-style-guide', '--extra-details']

    # Use Rubocop's Github Actions formatter if possible
    if ENV['GITHUB_ACTIONS'] == 'true'
      task.formatters << 'github'
    end
  end
end
```

IMHO the benefit of moving the actual code to `else` is that you better know where the exception can be triggered.

File: lib/beaker/dsl/structure.rb (hunk 6 above)
Comment (bastelfreak) [reply lib/beaker/dsl/structure.rb]: fixed now as well

File: .rubocop_todo.yml [hunk 14]
```diff
@@ -55,6 +75,11 @@ Lint/ShadowingOuterLocalVariable:
     - 'lib/beaker/host_prebuilt_steps.rb'
     - 'lib/beaker/perf.rb'
 
+# Configuration parameters: AllowComments, AllowNil.
+Lint/SuppressedException:
```
Comment (ekohl) [inline .rubocop_todo.yml:79]: I think you solved this, or are there more cases of it?

Comment (bastelfreak) [reply .rubocop_todo.yml]: done

**Review by ekohl — APPROVED** [review]
